libs/vkd3d/state.c

```diff
@@ -2536,16 +2538,25 @@ static HRESULT vkd3d_setup_shader_stage(struct d3d12_pipeline_state *state, stru
             required_subgroup_size_info->sType = VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_REQUIRED_SUBGROUP_SIZE_CREATE_INFO;
             required_subgroup_size_info->pNext = (void*)stage_desc->pNext;
 
-            if (spirv_code->meta.cs_required_wave_size)
+            /* [WaveSize(min,max,preferred)] in SM 6.8, or [WaveSize(N)] in SM 6.6. */
+            if (spirv_code->meta.cs_wave_size_preferred &&
+                    spirv_code->meta.cs_wave_size_preferred >= device->device_info.vulkan_1_3_properties.minSubgroupSize &&
+                    spirv_code->meta.cs_wave_size_preferred <= device->device_info.vulkan_1_3_properties.maxSubgroupSize)
             {
-                /* [WaveSize(N)] attribute in SM 6.6. */
-                subgroup_size_alignment = spirv_code->meta.cs_required_wave_size;
-
-                /* If min == max, we can still support WaveSize in a dummy kind of way. */
-                if (device->device_info.vulkan_1_3_properties.requiredSubgroupSizeStages & VK_SHADER_STAGE_COMPUTE_BIT)
-                    stage_desc->pNext = required_subgroup_size_info;
-
-                stage_desc->flags &= ~VK_PIPELINE_SHADER_STAGE_CREATE_ALLOW_VARYING_SUBGROUP_SIZE_BIT;
+                /* Preferred wave size is non-zero and is supported by the device. */
+                subgroup_size_alignment = spirv_code->meta.cs_wave_size_preferred;
+                override_subgroup_size = true;
+            }
+            else if (spirv_code->meta.cs_wave_size_min && (
+                    spirv_code->meta.cs_wave_size_min > device->device_info.vulkan_1_3_properties.minSubgroupSize ||
+                    spirv_code->meta.cs_wave_size_max < device->device_info.vulkan_1_3_properties.maxSubgroupSize))
+            {
+                /* We generally want to let the driver decide on a subgroup size, but if the device supports
+                 * subgroup sizes outside the range required by the shader, we need to specify it manually.
+                 * For now, prefer large subgroups in order to hit wave64 on AMD. */
```

Comment:
I would add a vendor check for Intel here and opt for smallest possible. Intel implements large wave in a slow way.

Reply:
Done.